ssi-sd-jwt/src/decode.rs

```diff
@@ -0,0 +1,235 @@
+
+    let sd_alg = sd_alg(&payload_claims)?;
+    let _ = payload_claims
+        .as_object_mut()
+        .unwrap()
+        .remove(SD_ALG_CLAIM_NAME);
+
+    let mut disclosures = translate_to_in_progress_disclosures(disclosures, sd_alg)?;
+
+    visit_claims(&mut payload_claims, &mut disclosures)?;
+
+    for (_, disclosure) in disclosures {
+        if !disclosure.found {
+            return Err(DecodeError::UnusedDisclosure);
+        }
+    }
+
+    Ok((validity_claims, serde_json::from_value(payload_claims)?))
+}
+
+fn sd_alg(claims: &serde_json::Value) -> Result<SdAlg, DecodeError> {
+    let alg_name = claims[SD_ALG_CLAIM_NAME]
+        .as_str()
+        .ok_or(DecodeError::MissingSdAlg)?;
+
+    SdAlg::try_from(alg_name)
+}
+
+fn translate_to_in_progress_disclosures(
+    disclosures: &[&str],
+    sd_alg: SdAlg,
+) -> Result<BTreeMap<String, InProgressDisclosure>, DecodeError> {
+    let disclosure_vec: Result<Vec<_>, DecodeError> = disclosures
+        .iter()
+        .map(|disclosure| InProgressDisclosure::new(disclosure, sd_alg))
+        .collect();
+
+    let disclosure_vec = disclosure_vec?;
+
+    let mut disclosure_map = BTreeMap::new();
+    for disclosure in disclosure_vec {
+        let prev = disclosure_map.insert(disclosure.hash.clone(), disclosure);
+
+        if prev.is_some() {
+            return Err(DecodeError::MultipleDisclosuresWithSameHash);
+        }
+    }
+
+    Ok(disclosure_map)
+}
+
+#[derive(Debug)]
+struct InProgressDisclosure {
+    decoded: DecodedDisclosure,
+    hash: String,
+    found: bool,
+}
+
+impl InProgressDisclosure {
+    fn new(disclosure: &str, sd_alg: SdAlg) -> Result<Self, DecodeError> {
+        Ok(InProgressDisclosure {
+            decoded: DecodedDisclosure::new(disclosure)?,
+            hash: hash_encoded_disclosure(sd_alg, disclosure),
+            found: false,
+        })
+    }
+}
+
+fn visit_claims(
+    payload_claims: &mut serde_json::Value,
+    disclosures: &mut BTreeMap<String, InProgressDisclosure>,
+) -> Result<(), DecodeError> {
+    let payload_claims = match payload_claims.as_object_mut() {
+        Some(obj) => obj,
+        None => return Ok(()),
+    };
+
+    // Visit children
+    for (_, child_claim) in payload_claims.iter_mut() {
+        visit_claims(child_claim, disclosures)?
+    }
+
+    // Process _sd claim
+    let new_claims = if let Some(sd_claims) = payload_claims.get(SD_CLAIM_NAME) {
+        decode_sd_claims(sd_claims, disclosures)?
+    } else {
+        vec![]
+    };
+
+    if payload_claims.contains_key(SD_CLAIM_NAME) {
+        payload_claims.remove(SD_CLAIM_NAME);
+    }
+
+    for (new_claim_name, mut new_claim_value) in new_claims {
+        visit_claims(&mut new_claim_value, disclosures)?;
+
+        let prev = payload_claims.insert(new_claim_name, new_claim_value);
+
+        if prev.is_some() {
+            return Err(DecodeError::DisclosureClaimCollidesWithJwtClaim);
```

Comment:
Consider warning instead of returning an error here.

Reply:
This is another requirement of the spec

> If the claim name already exists at the level of the _sd key, the SD-JWT MUST be rejected.